httpx/_multipart.py

```diff
@@ -78,23 +78,39 @@ def __init__(self, name: str, value: FileTypes) -> None:
 
         fileobj: FileContent
 
+        headers: typing.Dict[str, str] = {}
+        content_type: typing.Optional[str] = None
+
         if isinstance(value, tuple):
             try:
-                filename, fileobj, content_type = value  # type: ignore
+                filename, fileobj, content_type, headers = value  # type: ignore
             except ValueError:
-                filename, fileobj = value  # type: ignore
-                content_type = guess_content_type(filename)
+                try:
+                    filename, fileobj, content_type = value  # type: ignore
+                except ValueError:
+                    filename, fileobj = value  # type: ignore
+            else:
+                headers = {k.title(): v for k, v in headers.items()}
+                if "Content-Type" in headers:
+                    raise ValueError(
+                        "Content-Type cannot be included in multipart headers"
+                    )
```

Comment:
Perhaps we don't need to do this check.

It's *odd* behaviour for the developer to set the `content_type` and then override it with the *actual* value provided in the custom headers. But it's not *broken*.

My preference would be that we don't do the explicit check here. In the case of conflicts I'd probably have `header` values take precedence.

I'm not absolute on this one, but slight preference.

Reply:
Makes sense. I thought it'd be a good idea to check what requests does here. It looks like it silently ignores the header in the `header`. That is:

```python
requests.post("http://httpbin.org/post", files=[("test", ("test.txt", b"data", "text/plain", {"Content-Type": "text/csv"}))])
```

Gets sent as `text/plain`.

Digging into _why_ this is the case, it seems like it's just an implementation detail in urllib3. It happens [here](https://github.com/urllib3/urllib3/blob/c58dbbb35a19877a8f367c1eaa99ca43c2148f59/src/urllib3/fields.py#L351).

I'm not sure what the right thing to do here is, but if you feel like it's best to go with no error and making `header` values take precedence, I'm happy to implement that.

Reply:
> I'm not sure what the right thing to do here is, but if you feel like it's best to go with no error and making header values take precedence, I'm happy to implement that.

I reckon let's do that, yeah.

> Another alternative would be to have the 3rd parameter be either a string representing the content type or a headers dict. We can't really make the 3rd parameter always be a headers dict because that would be a breaking change for httpx.

I actually quite like that yes, neat idea. The big-tuples API is... not helpful really. But let's probably just go with the path of least resistance here. Perhaps one day we'll want an `httpx` 2.0, where we gradually start deprecating the various big-tuples bits of API in favour of a neater style.

Reply:
> I reckon let's do that, yeah.

👍 gonzo

> > Another alternative would be to have the 3rd parameter be either a string representing the content type or a headers dict. We can't really make the 3rd parameter always be a headers dict because that would be a breaking change for httpx.
> 
> I actually quite like that yes, neat idea. The big-tuples API is... not helpful really. But let's probably just go with the path of least resistance here. Perhaps one day we'll want an `httpx` 2.0, where we gradually start deprecating the various big-tuples bits of API in favour of a neater style.

Agreed! I added a comment in the code explaining the reasoning behind the big tuple API (inherited from requests) and how we might want to change it in the future.